libs/cart/src/reducers/cart-coupon/cart-coupon.reducer.ts

```diff
@@ -0,0 +1,80 @@
+import {
+  DaffCartCouponActionTypes,
+} from '../../actions/public_api';
+import { initialState } from '../cart-initial-state';
+import { DaffCartReducerState } from '../cart-state.interface';
+import { ActionTypes } from '../action-types.type';
+import { DaffCartErrorType } from '../cart-error-type.enum';
+import { DaffCart } from '../../models/cart';
+
+function addError<T extends DaffCart>(state: DaffCartReducerState<T>, error: string) {
+  return {
+    ...state,
+    errors: {
+      ...state.errors,
+      [DaffCartErrorType.Coupon]: state.errors[DaffCartErrorType.Coupon].concat(new Array(error))
+    }
+  };
+}
+
+function resetErrors<T extends DaffCart>(state: DaffCartReducerState<T>) {
+  return {
+    errors: {
+      ...state.errors,
+      [DaffCartErrorType.Coupon]: []
+    }
+  };
+}
+
+export function cartCouponReducer<T extends DaffCart>(
+  state = initialState,
+  action: ActionTypes
+): DaffCartReducerState<T> {
+  switch (action.type) {
+    case DaffCartCouponActionTypes.CartCouponApplyAction:
+    case DaffCartCouponActionTypes.CartCouponListAction:
+    case DaffCartCouponActionTypes.CartCouponRemoveAction:
+    case DaffCartCouponActionTypes.CartCouponRemoveAllAction:
+      return {
+        ...state,
+        loading: true
+      };
+
+    case DaffCartCouponActionTypes.CartCouponApplySuccessAction:
+    case DaffCartCouponActionTypes.CartCouponRemoveSuccessAction:
+    case DaffCartCouponActionTypes.CartCouponRemoveAllSuccessAction:
+      return {
+        ...state,
+        ...resetErrors(state),
+        cart: {
+          ...state.cart,
+          ...action.payload
```

Comment:
should this action return a `Partial<DaffCart>` instead of a `DaffCart`? Because won't this overwrite the fields not returned by the applyCoupon mutation? The applyCoupon mutation only requests a partial cart

Reply:
yeah good point, they should be partials

Reply:
That's my bad. When I was doing it originally, I thought that's how the other cart actions were done, but I must've misread them and didn't think it through this far

Reply:
No its all good, I overlooked that. The other cart actions use partials so I should have known.

Reply:
changed